Explorer/Assets/Plugins/TexturesFuse/TexturesServerWrap/Unzips/TexturesUnzip.cs

```diff
@@ -0,0 +1,339 @@
+            initOptions.outputMessage = debug ? OutputMessage : null;
+            var result = NativeMethods.TexturesFuseInitialize(initOptions, out context);
+
+            if (result is not NativeMethods.ImageResult.Success)
+                throw new Exception($"TexturesFuseInitialize failed: {result}");
+        }
+
+        [MonoPInvokeCallback(typeof(NativeMethods.OutputMessageDelegate))]
+        private static void OutputMessage(int format, string message)
+        {
+            ReportHub.Log(ReportCategory.TEXTURES, $"TexturesFuse: {message}");
+        }
+
+        private void ReleaseUnmanagedResources()
+        {
+            var result = NativeMethods.TexturesFuseDispose(context);
+
+            if (result is not NativeMethods.ImageResult.Success)
+                ReportHub.LogError(ReportCategory.TEXTURES, $"TexturesFuseDispose failed: {result}");
+        }
+
+        ~TexturesUnzip()
+        {
+            Dispose();
+        }
+
+        public void Dispose()
+        {
+            if (disposed)
+                return;
+
+            disposed = true;
+            ReleaseUnmanagedResources();
+            GC.SuppressFinalize(this);
+        }
+
+        public async UniTask<Result> TextureFromBytesAsync(
+            IntPtr bytes,
+            int bytesLength,
+            TextureType type,
+            CancellationToken token
+        )
+        {
+            if (Result.TryErrorIfCancelled(token, out var errorResult))
+                return errorResult;
+
+            await UniTask.SwitchToThreadPool();
+
+            if (Result.TryErrorIfCancelled(token, out errorResult))
+                return errorResult;
+
+            ProcessImage(bytes, bytesLength, type, out var handle, out var pointer, out int outputLength, out NativeMethods.ImageResult result, out uint width, out uint height, out bool linear, out TextureFormat format);
+            await UniTask.SwitchToMainThread();
+
+            if (result is NativeMethods.ImageResult.Success && token.IsCancellationRequested)
+            {
+                NativeMethods.TexturesFuseRelease(context, handle);
+                if (Result.TryErrorIfCancelled(token, out errorResult))
+                    return errorResult;
+            }
+
+            if (result is not NativeMethods.ImageResult.Success)
+            {
+                ReportHub.LogError(ReportCategory.TEXTURES, $"TexturesFuseASTCImageFromMemory error during decoding: {result}");
+                return EnumResult<OwnedTexture2D, NativeMethods.ImageResult>.ErrorResult(result, string.Empty);
+            }
+
+            if (handle == IntPtr.Zero)
+                throw new Exception("TexturesFuseProcessedImageFromMemory failed");
+
+            //TODO do linear on RGBA?
+            //TODO mipChain and mipmaps
+            var texture = new Texture2D((int)width, (int)height, format, false, linear, true);
+            texture.LoadRawTextureData(pointer, outputLength);
```

Comment:
Is it fine to execute it on the main thread?

Reply:
I assumed that LoadRawTextureData is called only from main thread since it's Unity's API

Reply:
In theory you could use [GetRawTextureData](https://docs.unity3d.com/ScriptReference/Texture2D.GetRawTextureData.html), switch to the background thread to write data into it, and then call `Apply` on the main thread, but I am not sure if it will bring any benefits. 
It requires a bit of profiling, e.g. with stopwatch.

If the current way does not produce bottle-neck with big textures I would not bother